lib/improver/nbhood/use_nbhood.py

```diff
@@ -156,3 +159,133 @@ def process(self, cube, mask_cube):
+
+
+class CollapseMaskedNeighbourhoodCoordinate(object):
+
+    """
+    Plugin for collapsing the coordinate the mask was applied to after
+    masked neighbourhood processing.
+
+    Takes into account the result from the neighbourhood processing to
+    adjust the weights between the bands in the coordinate for the points
+    where the were no points within a neighbourhood for and a non-zero
+    weighting.
+    """
+
+    def __init__(self, coord_masked, weights=None):
+        """
+        Initialise the class.
+
+        Args:
+            coord_masked (string):
+                String matching the name of the coordinate that has been used
+                for masking.
+
+        Keyword Args:
+            weights (cube):
+                A cube from an ancillary file containing the weights for each
+                point in the coord_masked at each grid point. Only two points
+                in coord_masked can have a non-zero weight for each grid-point,
+                i.e. we are only weighting between two adjacent bands in the
+                neighbourhood output for each gridpoint.
+                Should have the coordinates coord_masked, x and y.
+                Default is None which equal weights for each band in the mean
+                used to collapse the chosen coordinate.
+                The weights cube can be masked, and this mask will be retained,
+                and will be present in the output.
+.
+        """
+        self.coord_masked = coord_masked
+        self.weights = weights
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<ApplyNeighbourhoodProcessingWithAMask: '
+                  'coord_masked: {}, weights: {}>')
+        return result.format(self.coord_masked,
+                             self.weights)
+
+    def reweight_weights(self, nbhood_cube):
+        """
+        Reweight the weights taking into account where there are NaNs in the
+        result from neighbourhood.
+
+        The weights corresponding to NaNs in the result from neighbourhooding
+        with a mask are set to zero and then the weights are renormalized along
+        the axis corresponding to the coordinate we want to collapse.
+
+        Args:
+            nbhood_cube (iris.cube.Cube):
+                The cube that has been through masked neighbourhood processing
+                and has the dimension we wish to collapse. Must have the same
+                dimensions of the cube.
+        """
+        # If the weights are masked we want to retain the mask.
+        condition = np.isnan(nbhood_cube.data)
+        if ma.is_masked(self.weights.data):
+            condition = condition & ~self.weights.data.mask
+
+        self.weights.data[condition] = 0.0
+        axis = nbhood_cube.coord_dims(self.coord_masked)
+        self.weights.data = WeightsUtilities.normalise_weights(
+            self.weights.data, axis=axis)
+
+    def process(self, cube):
```

Comment:
Having looked at the unit tests, I'm feeling that this function does a bit too much and could be improved by breaking down a little more.  For example, a separate function to collapse the cube (lines 273-279) for which tests would then check only the output_cube.data; and one to remove the collapsed coordinate and methods (lines 280-290), for which tests would cover that area.  This would allow the unit tests (and particularly some of the setting up methods) to be clarified significantly.

Reply:
I've factored out the metadata manipulation stuff which I think reduces some of the complexity. I don't really know how much further to go so it might be worth a chat about it.